src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceValueSerializer.cs

```diff
@@ -0,0 +1,222 @@
+            }
+
+            IEdmTypeReference edmType = writeContext.GetEdmType(graph, type);
+            Contract.Assert(edmType != null);
+
+            messageWriter.WriteProperty(CreateProperty(graph, edmType, writeContext.RootElementName, writeContext));
+        }
+
+        /// <inheritdoc/>
+        public sealed override ODataValue CreateODataValue(object graph, IEdmTypeReference expectedType, ODataSerializerContext writeContext)
+        {
+            if (!expectedType.IsStructured())
+            {
+                throw Error.InvalidOperation(SRResources.CannotWriteType, typeof(ODataResourceValueSerializer), expectedType.FullName());
+            }
+
+            if (graph == null)
+            {
+                return new ODataNullValue();
+            }
+
+            ODataResourceValue value = CreateODataResourceValue(graph, expectedType as IEdmStructuredTypeReference, writeContext);
+            if (value == null)
+            {
+                return new ODataNullValue();
+            }
+
+            return value;
+        }
+
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Performance", "CA1804:RemoveUnusedLocals", MessageId = "edmTypeSerializer")]
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Performance", "CA1822:MarkMembersAsStatic")]
+        private ODataResourceValue CreateODataResourceValue(object graph, IEdmStructuredTypeReference expectedType, ODataSerializerContext writeContext)
+        {
+            List<ODataProperty> properties = new List<ODataProperty>();
+            ODataResourceValue resourceValue = new ODataResourceValue { TypeName = writeContext.GetEdmType(graph, graph.GetType()).FullName() };
+  
+            IDelta delta = graph as IDelta;
+            if (delta != null)
+            {
+                foreach (string propertyName in delta.GetChangedPropertyNames())
+                {
+                    object propertyValue;
+                    
+                    if (delta.TryGetPropertyValue(propertyName, out propertyValue))
+                    {
+                        Type propertyType;
+                        IEdmStructuredTypeReference expectedPropType = null;
+
+                        if (propertyValue == null)
+                        {
+                            // We need expected property type only if property value is null, else it will get from the value
+                            if (delta.TryGetPropertyType(propertyName, out propertyType))
+                            {
+                                expectedPropType = writeContext.GetEdmType(propertyValue, propertyType) as IEdmStructuredTypeReference;
+                            }                            
+                        }
+                        
+                        SetPropertyValue(writeContext, properties, expectedPropType, propertyName, propertyValue);
+                    }
```

Comment:
This logic is repeated in lines 125-141; would be better to create a common function called in these two places.